src/storage/core.rs

```diff
@@ -203,7 +203,25 @@ impl<K> Storage<K> {
             .active_blob
             .as_mut()
             .ok_or_else(Error::active_blob_not_set)?;
-        blob.write(record).await
+        match blob.write(record).await {
+            Ok(_) => Ok(()),
+            Err(err) => match err.downcast::<Error>() {
```

Comment:
I think you can reduce nesting here, with something like that: 
```
     blob.write(record).await.or_else(|e| {
            let err = e.downcast::<Error>()?;

            if let ErrorKind::FileUnavailable(kind) = err.kind() {
                let dir = self.inner.config.work_dir().ok_or(Error::uninitialized())?;
                return Err(Error::work_dir_unavailable(dir, err.to_string(), *kind).into());
            }

            Err(err.into())
        })

```
Not sure about using "if let return" or "match", though

Reply:
Probably it's even better to completely eliminate multiple return points by making err mutable and assigning it in the branch